flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/Solver.scala

```diff
@@ -0,0 +1,146 @@
+  // TODO(tvas): Maybe we want to pass a WeightVector directly here, instead of a
+  // DataSet[WeightVector]
+
+  /** Creates a DataSet with one zero vector. The zero vector has dimension d, which is given
+    * by the dimensionDS.
+    *
+    * @param dimensionDS DataSet with one element d, denoting the dimension of the returned zero
+    *                    vector
+    * @return DataSet of a zero vector of dimension d
+    */
+  def createInitialWeightVector(dimensionDS: DataSet[Int]):  DataSet[WeightVector] = {
+    dimensionDS.map {
+      dimension =>
+        val values = Array.fill(dimension)(0.0)
+        new WeightVector(DenseVector(values), 0.0)
+    }
+  }
+
+  //Setters for parameters
+  def setLossFunction(lossFunction: LossFunction): Solver = {
+    parameters.add(LossFunction, lossFunction)
+    this
+  }
+
+  def setRegularizationType(regularization: RegularizationType): Solver = {
+    parameters.add(RegularizationType, regularization)
+    this
+  }
+
+  def setRegularizationParameter(regularizationParameter: Double): Solver = {
+    parameters.add(RegularizationParameter, regularizationParameter)
+    this
+  }
+
+  def setPredictionFunction(predictionFunction: PredictionFunction): Solver = {
+    parameters.add(PredictionFunctionParam, predictionFunction)
+    this
+  }
+}
+
+object Solver {
+  // TODO(tvas): Does this belong in IterativeSolver instead?
+  val WEIGHTVECTOR_BROADCAST = "weights_broadcast"
+
+  // Define parameters for Solver
+  case object LossFunction extends Parameter[LossFunction] {
+    // TODO(tvas): Should depend on problem, here is where differentiating between classification
+    // and regression could become useful
+    val defaultValue = Some(new SquaredLoss)
+  }
+
+  case object RegularizationType extends Parameter[RegularizationType] {
+    val defaultValue = Some(new NoRegularization)
+  }
+
+  case object RegularizationParameter extends Parameter[Double] {
+    val defaultValue = Some(0.0) // TODO(tvas): Properly initialize this, ensure Parameter > 0!
+  }
+
+  case object PredictionFunctionParam extends Parameter[PredictionFunction] {
+    val defaultValue = Some(new LinearPrediction)
+  }
+}
+
+/** An abstract class for iterative optimization algorithms
+  *
+  * See [[https://en.wikipedia.org/wiki/Iterative_method Iterative Methods on Wikipedia]] for more
+  * info
+  *
+  */
+abstract class IterativeSolver extends Solver {
+
+  //Setters for parameters
+  def setIterations(iterations: Int): IterativeSolver = {
+    parameters.add(Iterations, iterations)
+    this
+  }
+
+  def setStepsize(stepsize: Double): IterativeSolver = {
+    parameters.add(Stepsize, stepsize)
+    this
+  }
+
+  //  def setConvergenceThreshold(convergenceThreshold: Double): GradientDescent = {
+  //    parameters.add(ConvergenceThreshold, convergenceThreshold)
+  //    this
+  //  }
+
+  //  case object ConvergenceThreshold extends Parameter[Double] {
+  //    val defaultValue = None
+  //  }
```

Comment:
Commented code should not be contained. Once we implement the convergence threshold, we can re-add this code.